src/main/java/com/datadoghq/sketch/ddsketch/store/PaginatedStore.java

```diff
@@ -0,0 +1,387 @@
+                        if (pages[i][j] != 0D) {
+                            return ((i + minPageIndex) << PAGE_SHIFT) + j;
+                        }
+                    }
+                }
+            }
+        }
+        throw new NoSuchElementException();
+    }
+
+    @Override
+    public int getMaxIndex() {
+        if (null != pages) {
+            for (int i = pages.length - 1; i >= 0; --i) {
+                if (null != pages[i]) {
+                    for (int j = pages[i].length - 1; j >= 0; --j) {
+                        if (pages[i][j] != 0D) {
+                            return ((i + minPageIndex) << PAGE_SHIFT) + j;
+                        }
+                    }
+                }
+            }
+        }
+        throw new NoSuchElementException();
+    }
+
+    @Override
+    public double getTotalCount() {
+        if (isEmpty()) {
+            return 0D;
+        }
+        double total = 0D;
+        for (double[] page : pages) {
+            if (null != page) {
+                for (double count : page) {
+                    total += count;
+                }
+            }
+        }
+        return total;
+    }
+
+    @Override
+    public void add(int index, double count) {
+        if (count > 0) {
+            int alignedIndex = alignedIndex(index);
+            double[] page = getPage(alignedIndex >>> PAGE_SHIFT);
+            page[alignedIndex & PAGE_MASK] += count;
+        }
+    }
+
+    private double[] getPage(int pageIndex) {
+        double[] page = pages[pageIndex];
+        if (null == page) {
+            page = pages[pageIndex] = new double[PAGE_SIZE];
+        }
+        return page;
+    }
+
+    private int alignedIndex(int index) {
+        // get the index of the page this value should be stored in
+        int pageIndex = index < 0
+                ? ~(-index >>> PAGE_SHIFT)
+                : index >>> PAGE_SHIFT;
+        if (pageIndex < minPageIndex) {
+            // then space needs to be made before the first page,
+            // unless this is the first insertion
+            if (null == pages) {
+                lazyInit(pageIndex);
+            } else {
+                shiftPagesRight(pageIndex);
+            }
+        } else if (pageIndex >= minPageIndex + pages.length - 1) {
+            // then space needs to be made after the last page
+            extendTo(pageIndex);
+        }
+        // align the index relative to the start of the sketch
+        return index + (-minPageIndex << PAGE_SHIFT);
+    }
+
+    private void lazyInit(int pageIndex) {
+        minPageIndex = pageIndex;
+        pages = new double[GROWTH][];
+    }
+
+    private void shiftPagesRight(int pageIndex) {
+        int requiredExtension = minPageIndex - pageIndex;
+        // check if there is space to shift into
+        boolean canShiftRight = true;
+        for (int i = 0; i < requiredExtension && canShiftRight; ++i) {
+            canShiftRight = null == pages[pages.length - i - 1];
```

Comment:
Had to do read this twice, but I assume that even if `requiredExtension` is bigger than `pages.length`, we can never go out of bounds in the array, since there must be at least one page in the array, and then `canShiftRight` will become `false`.

Reply:
Exactly, but needing to read it twice makes me think I need to make it clearer.

Reply:
I think a comment about the invariants would be sufficient.